doc/adr/0016- end-to-end testing-strategy.rst

```diff
@@ -0,0 +1,199 @@
+
+Although ``cardano-ledger-api`` is under test it isn't feasible to expect thorough coverage of all
+node functionality, such as staking and update proposals, because the primary focus is to satisify
+end-to-end testing requirements for ``plutus``. Fortunately, much of the node and ledger functionality
+is already covered by ``cardano-node-tests``.
+
+Initially, a few tests that do not depend on `plutus-apps
+<https://github.com/input-output-hk/plutus-apps/>`_  will be created. This entails building
+transactions with ``cardano-ledger-api`` and waiting for on-chain events using ``cardano-node-client``
+without use of the constraints library. This approach allows us to build transactions specifically,
+which is especially useful for testing edge-cases and error scenarios. Although, it will likely
+require more boilerplate code and negatively impact readabity.
+
+We may decide early on to depend on ``plutus-apps`` to use the Contract monad, which would give a
+uniform interface for off-chain code such as different node backends (private and public testnets,
+and emulator) and chain-indexer queryies (``cardano-node-client`` or Marconi in future). It should
+also reduce the amount of boilerplate code and provide additional features such as trace logging.
+
+If we wish to use the ``cardano-node-emulator``, which is a faster node without consensus, for
+property based testing that requires faster test execution then we'll need to depend on ``plutus-apps``.
+Alternatively, those types of tests would exist only in ``plutus-apps``. 
+
+There is the option of including additional packages to test from ``plutus-apps`` at a later stage.
+
+Types of tests for the Haskell framework
+----------------------------------------
+All ``plutus`` end-to-end testing requirements will be covered by the Haskell framework. Arguments for
+and against this approach are covered in the "Argument" section below. In summary, with access to
+the Haskell and Plutus interfaces and reduced friction from using a single programming language we
+are likely improve test coverage at this level. E.g. builtin functions and error scenarios.
+
+Although we will be building out the end-to-end test covereage, it is more efficient to have fewer
+and broader test scenarios at this level and a greater number of tests at the lower unit and
+integration levels for stressing particular features and covering negative scenarios and edge cases.
+
+Examples of types of tests to be included
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Any Plutus Core builtin function. These may already be tested extensively in the lower 
+  unit/property/intgration levels but there's value in having some coverage at the end-to-end
+  level too.
+* Use cases that go beyond testing features in isolation. Bringing together various functionality
+  helps demonstrate the capability of more realistic plutus applications.
+* Latest Plutus version, which would likely introduce new changes to the cardano ledger. Perhaps
```

Comment:
I recall you saying the test framework would be for **all** plutus versions, no? Saing "Latest Plutus version" might make us think that you want to limit the scope of testing framework to the latest Plutus version.